processor/processor_bot_enricher_test.go

```diff
@@ -0,0 +1,277 @@
+				require.False(t, gjson.Get(event, "context.bot.url").Exists(), "bot URL should not be present when invalid browser")
+			})
+	})
+}
+
+type botScenario struct {
+	botEnrichmentEnabled bool
+	isBot                bool
+	botName              string
+	botURL               string
+	isInvalidBrowser     bool
+}
+
+func (s *botScenario) WithBotEnrichmentEnabled(enabled bool) *botScenario {
+	s.botEnrichmentEnabled = enabled
+	return s
+}
+
+func (s *botScenario) WithBotInfo(isBot bool, name, url string, isInvalidBrowser bool) *botScenario {
+	s.isBot = isBot
+	s.botName = name
+	s.botURL = url
+	s.isInvalidBrowser = isInvalidBrowser
+	return s
+}
+
+func (s *botScenario) Run(t *testing.T, verification func(t *testing.T, event string)) {
+	config.Reset()
+	defer config.Reset()
+	writeKey := "writekey-1"
+	workspaceID := "workspace-1"
+	sourceID := "source-1"
+
+	gatewayUrl, db, cancel, wg := s.startAll(t, writeKey, workspaceID, sourceID)
+	defer func() {
+		cancel()
+		_ = wg.Wait()
+	}()
+	require.NoError(t, s.sendEvent(gatewayUrl, writeKey, workspaceID, sourceID))
+
+	s.requireJobsCount(t, db, "gw", "succeeded", 1)
+	s.requireJobsCount(t, db, "rt", "aborted", 1)
+	var payload string
+	require.NoError(t, db.QueryRow("SELECT event_payload FROM rt_jobs_1").Scan(&payload))
+	verification(t, payload)
+}
+
+func (s *botScenario) startAll(t *testing.T, writeKey, workspaceID, sourceID string) (gatewayUrl string, db *sql.DB, cancel context.CancelFunc, wg *errgroup.Group) {
+	ctx, cancel := context.WithCancel(context.Background())
+	wg, ctx = errgroup.WithContext(ctx)
+
+	bcserver := backendconfigtest.NewBuilder().
+		WithWorkspaceConfig(
+			backendconfigtest.NewConfigBuilder().
+				WithWorkspaceID(workspaceID).
+				WithSource(
+					backendconfigtest.NewSourceBuilder().
+						WithID(sourceID).
+						WithWriteKey(writeKey).
+						WithConnection(
+							backendconfigtest.NewDestinationBuilder("WEBHOOK").
+								WithID("destination-1").
+								Build()).
+						Build()).
+				Build()).
+		Build()
+
+	trServer := transformertest.NewBuilder().Build()
+
+	pool, err := dockertest.NewPool("")
+	require.NoError(t, err)
+	postgresContainer, err := postgres.Setup(pool, t)
+	require.NoError(t, err)
+
+	gwPort, err := kithelper.GetFreePort()
+	require.NoError(t, err)
+	wg.Go(func() error {
+		err := s.runRudderServer(ctx, gwPort, postgresContainer, bcserver.URL, trServer.URL, t.TempDir())
+		if err != nil {
+			t.Logf("rudder-server exited with error: %v", err)
+		} else {
+			t.Log("rudder-server stopped")
+		}
+		bcserver.Close()
+		trServer.Close()
+		return err
+	})
+	url := fmt.Sprintf("http://localhost:%d", gwPort)
+	health.WaitUntilReady(ctx, t, url+"/health", 60*time.Second, 10*time.Millisecond, t.Name())
+	return url, postgresContainer.DB, cancel, wg
+}
+
+func (s *botScenario) runRudderServer(ctx context.Context, port int, postgresContainer *postgres.Resource, cbURL, transformerURL, tmpDir string) (err error) {
+	config.Set("enableStats", false)
```

Comment:
It's a shame that we are still using global variables in our test. I understand that refactoring the runner to make it use a `*config.Config` instance is beyond the scope of this PR, just making a note of this.

You can leave it as is or use `t.Setenv`. Sooner or later we should remove this as it could lead to undesired behaviour cc @ktgowtham

Reply:
This is also preventing us to use `t.Parallel()`.

Reply:
https://linear.app/rudderstack/issue/PIPE-2067/get-rid-of-global-variables-in-tests